Updating Direct ML Execution Provider - Issue 10324 (#11554)
microsoft/onnxruntime: docs/execution-providers/DirectML-ExecutionProvider.md

The high level design looks like this:

![ONNX + WinML layered architecture](../../images/layered-architecture.png)

### API choice
{: .no_toc }

You can choose to use either the WinRT API or the C API.

||WinRT|C API|
|--|--|--|
|Type system| Integration with Windows RT types| Platform neutral types|
|Language support| Language support via WinRT Projections| Language support via per language projections|
|Tensorization| Accepts VideoFrames and converts to tensors (support for CPU and GPU)| Accepts tensors|

### Using the NuGet WinRT API with other C-API distributions
{: .no_toc }

The WinRT API NuGet package is distributed with a specific version of ONNX Runtime, but apps can include their own version of ONNX Runtime (either a [released version](../install/#cccwinml-installs) or [a custom build](../build/)). You may wish to do this to use non-default execution providers.
To use your own version of ONNX Runtime, replace onnxruntime.dll with your desired version.

<p><a href="#">Back to top</a></p>

## Install
Pre-built packages of ORT with the DirectML EP is published on Nuget.org. See: [Install ORT](../install#cccwinml-installs).

## Requirements

> The [DirectML](https://github.com/microsoft/DirectML) execution provider requires a [DirectX 12](https://answers.microsoft.com/en-us/windows/forum/all/direct-x-120-download/98d1137a-653d-40eb-9863-3b474665add7) capable device. Almost all commercially-available graphics cards released in the last several years support DirectX 12. Here are some examples of compatible hardware:

Reviewer: No need for a link from DirectML, since it was already used in the first paragraph of this page and it has a link there.

The link on `DirectX 12` doesn't seem helpful to me. I'd remove it.

* NVIDIA Kepler (GTX 600 series) and above
* AMD GCN 1st Gen (Radeon HD 7000 series) and above
* Intel Haswell (4th-gen core) HD Integrated Graphics and above
* Qualcomm Adreno 600 and above

> [DirectML](https://docs.microsoft.com/en-us/windows/ai/directml/dml) is introduced in Windows 10, version 1903, and in the corresponding version of the Windows SDK.

Reviewer: No need for a link from DirectML, since it was already used in the first paragraph of this page and it has a link there.

`is introduced` -> `was introduced` seems more grammatically correct to me.

<p><a href="#">Back to top</a></p>

## Build

Requirements for building the DirectML execution provider:
1. Visual Studio 2017 toolchain
2. [The Windows 10 SDK (10.0.18362.0) for Windows 10, version 1903](https://developer.microsoft.com/en-us/windows/downloads/windows-10-sdk) (or newer)

> To build onnxruntime with the DML EP included, supply the `--use_dml` parameter to `build.bat`. 

Reviewer: ```suggestion
To build onnxruntime with the DML EP included, supply the `--use_dml` flag to `build.bat`. 
```

Additionally, as the DirectML execution provider does not support parallel execution, it does not support multi-threaded calls to `Run` on the same inference session. That is, if an inference session using the DirectML execution provider, only one thread may call `Run` at a time. Multiple threads are permitted to call `Run` simultaneously if they operate on different inference session objects.

<p><a href="#">Back to top</a></p>

## Performance Tuning

The DirectML execution provider works most efficiently when tensor shapes are known at the time a session is created.  Here are a few performance benefits:

> 1. Constant folding can occur more often, there by reducing the CPU / GPU copies and stalls during evaluations.

Reviewer: ```suggestion
1. Constant folding can occur more often, reducing the CPU / GPU copies and stalls during evaluations.
```

Normally, when the shapes of model inputs are known during a session creation, the shapes for the rest of the model are inferred by OnnxRuntime when that session is created.  

However, if a model input contains a free dimension (such as for batch size), additional steps must be taken to retain the above performance benefits. These include:

1. Edit the model to replace an input's free dimension (specified through ONNX using "dim_param") with a fixed size (specified through ONNX using "dim_value").
2. Specify values of named dimensions within model inputs when creating the session using the OnnxRuntime *AddFreeDimensionOverrideByName* ABI.
3. Edit the model to ensure that an input's free dimension has a [denotation](https://github.com/onnx/onnx/blob/master/docs/DimensionDenotation.md) (such as "DATA_BATCH," or a custom denotation).  Then, when creating the session, specify the dimension size for each denotation. This can be done using the OnnxRuntime *AddFreeDimensionOverride* ABI.

<p><a href="#">Back to top</a></p>

## Samples

> DirectML and Onnx Runtime can be used to achieve many models. Here are some examples given on [Github](https://github.com/microsoft/DirectML#directml-samples):

Reviewer: Copying the list of mdoels from that page to this one seems worse to me than just leaving a link here.

> DirectML and Onnx Runtime can be used to achieve many models. Here are some examples given on [Github](https://github.com/microsoft/DirectML#directml-samples):

Reviewer: ```suggestion
[The DirectML GitHub repo](https://github.com/microsoft/DirectML/blob/master/README.md#directml-samples) has sample models which can be executed by the DirectML Execution Provider.
```

- yolov4: YOLOv4 is an object detection model capable of recognizing up to 80 different classes of objects in an image. This sample contains a complete end-to-end implementation of the model using DirectML, and is able to run in real time on a user-provided video stream.
- MobileNet: Adapted from the ONNX MobileNet model. MobileNet classifies an image into 1000 different classes. It is highly efficient in speed and size, ideal for mobile applications.
- MNIST: Adapted from the ONNX MNIST model. MNIST predicts handwritten digits using a convolution neural network.
- SqueezeNet: Based on the ONNX SqueezeNet model. SqueezeNet performs image classification trained on the ImageNet dataset. It is highly efficient and provides results with good accuracy.
- FNS-Candy: Adapted from the Windows ML Style Transfer model sample, FNS-Candy re-applies specific artistic styles on regular images.
- Super Resolution: Adapted from the ONNX Super Resolution model, Super-Res upscales and sharpens the input images to refine the details and improve image quality.

> A complete sample of onnxruntime using the DirectML execution provider can be found under [samples/c_cxx/fns_candy_style_transfer](https://github.com/microsoft/onnxruntime/tree/master/samples//c_cxx/fns_candy_style_transfer).

Reviewer: this link seems broken
